File: js/lib/MessageBroker/dispatchers/RabbitMQ/routings/Explicit/ExplicitRabbitMQTopic.js
const BaseRabbitMQTopic = require("../BaseRabbitMQTopic");
const ExplicitRabbitMQRoom = require("./ExplicitRabbitMQRoom");

class ExplicitRabbitMQTopic extends BaseRabbitMQTopic {
  constructor({ name, channel }) {
    super();

    this.name = name;
    this.channel = channel;
    this.channel.assertExchange(name, 'direct', {durable: false});
  }

  async createRoom({ name }) {
    let room = await ExplicitRabbitMQRoom.create({ 
      name, 
      topicName: this.name, 
      channel: this.channel 
    });

    this._addRoom(room);

    return room;
  }

  publish({ room, payload, options }) {
    this.channel.publish(this.name, room, new Buffer(payload), options);
  }
}

module.exports = ExplicitRabbitMQTopic

File: ruby/lib/MessageBroker/dispatchers/RabbitMQ/routings/Explicit/ExplicitRabbitMQTopic.rb
require_relative '../BaseRabbitMQTopic'
require_relative './ExplicitRabbitMQRoom'

class ExplicitRabbitMQTopic < BaseRabbitMQTopic
  def initialize name:, channel:
    super()

    @channel = channel
    @exchange = @channel.direct(name)
  end 

  def createRoom name:
    room = ExplicitRabbitMQRoom.new(
      name: name,
      channel: @channel,
      exchange: @exchange
    )
    
    addRoom(room: room)

    return room
  end

  def publish room:, payload:, correlationId: nil, replyTo: nil
    @exchange.publish(
      payload, 
      :routing_key => room, 
      :correlation_id => correlationId,
      :replyTo => replyTo
    )
  end
end